idea/data-storage/src/gear/gear-event-listener.ts

```diff
@@ -0,0 +1,277 @@
+
+    await this.handleMessageExtrinsics(block, status, timestamp);
+  }
+
+  private async handleMessageExtrinsics(block: SignedBlockExtended, status: ExtrinsicStatus, timestamp: number) {
+    const txMethods = ['sendMessage', 'sendReply', 'uploadProgram', 'createProgram'];
+    const extrinsics = block.block.extrinsics.filter(({ method: { method } }) => txMethods.includes(method));
+    const messages: Message[] = [];
+
+    if (extrinsics.length >= 1) {
+      for (const tx of extrinsics) {
+        const {
+          data: { id, source, destination, entry },
+        } = filterEvents(tx.hash, block, block.events, status).events.find(({ event }) =>
+          this.api.events.gear.MessageEnqueued.is(event),
+        ).event as MessageEnqueued;
+
+        const [payload, value] = getPayloadAndValue(tx.args, tx.method.method);
+
+        messages.push(
+          plainToClass(Message, {
+            id: id.toHex(),
+            destination: destination.toHex(),
+            source: source.toHex(),
+            payload,
+            value,
+            timestamp: new Date(timestamp),
+            genesis: this.genesis,
+            program: await this.programRepository.get(destination.toHex(), this.genesis),
+            type: MessageType.ENQUEUED,
+            entry: entry.isInit
+              ? MessageEntryPoint.INIT
+              : entry.isHandle
+                ? MessageEntryPoint.HANDLE
+                : MessageEntryPoint.REPLY,
+          }),
+        );
+      }
+    }
+
+    if (messages.length >= 1) await this.messageService.createMessages(messages);
+  }
+
+  private async handleProgramExtrinsics(block: SignedBlockExtended, status: ExtrinsicStatus, timestamp: number) {
+    const txMethods = ['uploadProgram', 'createProgram'];
+    const extrinsics = block.block.extrinsics.filter(({ method: { method } }) => txMethods.includes(method));
+    const programs: CreateProgramInput[] = [];
+
+    if (extrinsics.length >= 1) {
+      for (const tx of extrinsics) {
+        const {
+          data: { source, destination }
+        } = filterEvents(tx.hash, block, block.events, status).events.find(({ event }) =>
+          this.api.events.gear.MessageEnqueued.is(event),
+        ).event as MessageEnqueued;
+
+        let codeId;
+        const uploadProgramIndex = 0;
+        const createProgramIndex = 1;
+
+        if(tx.method.method === txMethods[uploadProgramIndex]) codeId = generateCodeHash(tx.args[0].toHex());
+        if(tx.method.method === txMethods[createProgramIndex]) codeId = tx.args[0].toHex();
+
+        programs.push({
+          owner: source.toHex(),
+          id: destination.toHex(),
+          blockHash: block.createdAtHash.toHex(),
+          timestamp,
+          code: await this.codeRepository.get(codeId, this.genesis),
+          genesis: this.genesis,
+        });
+      }
+    }
+
+    if (programs.length >= 1) await this.programService.createPrograms(programs);
```

Comment:
Is there a case where `programs.length` can be 0? It's already checked that `extrinsic.length` > 0, in this case you can be sure that programs won't be equal 0. But you need to move it to the previous block

Reply:
I get what you mean. Will update all three handlers.